File: CPP/No725.cc
/**
 * Created by Xiaozhong on 2020/11/26.
 * Copyright (c) 2020/11/26 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    vector<ListNode *> splitListToParts(ListNode *root, int k) {
        vector<ListNode *> ans;
        ListNode *cur = root;
        int n = 0;
        // 首先统计有多少个节点
        while (cur) {
            ++n;
            cur = cur->next;
        }
        // width 代表每一个新的链表块至少有多少个元素
        // remainder 代表从前往后有多少个元素需要多放一个元素进来
        int width = n / k, remainder = n % k;
        cur = root;
        // 从前往后挨个创建
        for (int i = 0; i < k; ++i) {
            // head 是伪头部节点，writer 是用来更新节点的
            ListNode *head = new ListNode(0), *writer = head;
            for (int j = 0; j < width + (i < remainder ? 1 : 0); ++j) {
                writer = writer->next = new ListNode(cur->val);
                if (cur != nullptr) cur = cur->next;
            }
            ans.emplace_back(head->next);
        }
        return ans;
    }
};

File: Java/src/No725.java
/*
 * Copyright (c) 9/22/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.Arrays;

public class No725 {
    static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode[] splitListToParts(ListNode head, int k) {
        int size = 0;
        ListNode curr = head;
        while (curr != null) {
            size++;
            curr = curr.next;
        }
        ListNode[] ans = new ListNode[k];
        if (size == 0) return ans;
        curr = head;
        ListNode prev = null;
        int every = size / k, remain = size % k;
        for (int i = 0; i < k; ++i) {
            ans[i] = curr;
            int tEvery = every, tRemain = remain;
            while (tEvery-- > 0) {
                prev = curr;
                curr = curr.next;
            }
            if (tRemain > i) {
                prev = curr;
                curr = curr.next;
            }
            prev.next = null;
        }
        return ans;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(1), l2 = new ListNode(2), l3 = new ListNode(3);
        l1.next = l2;
        l2.next = l3;
        No725 no725 = new No725();
        System.out.println(Arrays.toString(no725.splitListToParts(l1, 5)));
    }
}
